### ISSUE TITLE
Question about metadynamics implementation

### ISSUE THREAD

**Issue by ajsilveira** [openmm/openmm#2604]
Hi @peastman . I have a question regarding the implementation of well-tempered metadynamics.
When adding a gaussian, the collective variable is normalized. However, 
biasWidth enters the equation https://github.com/openmm/openmm/blob/2e86184b4b720f267d78d6c2902fff908695c135/wrappers/python/simtk/openmm/app/metadynamics.py#L199
without being divided by (maxValue - minValue). Should the user provide a normalized 
standard deviation? Also, I don't understand the role of gridWidth in that equation. 
Could you please provide more details about this? Thanks!

**Comment by craabreu** [comment]
Hi all,
There is another possible issue with `metadynamics.py`. While deposited Gaussians are correctly handled in the case of a periodic `BiasVariable`, nothing is passed to `Continuous1DFunction` objects (or its 2D/3D counterparts) with respect to periodic boundaries.
This could be easily solved for 1D functions because periodic splines are already available here:
https://github.com/openmm/openmm/blob/f9106ddb6a0d72c960c73b23ed0e07cd7190a80f/openmmapi/src/SplineFitter.cpp#L77
However, periodic 2D/3D splines would require further implementation.
@peastman, am I missing something here?

**Comment by craabreu** [comment]
I tested `Metadynamics` by making it deposit a single Gaussian and comparing the result with the exact expression. Please see the script and the graphical comparison below. The employed PDB file is [here](https://github.com/openmm/openmm/blob/master/wrappers/python/tests/systems/alanine-dipeptide-implicit.pdb). Clearly, the deposited Gaussian does not have the specified standard deviation. 

```
from simtk import openmm
from simtk.openmm import app
import matplotlib.pyplot as plt
import numpy as np

npoints = 181
sigma = 0.35
nsteps = 500

pdb = app.PDBFile('alanine-dipeptide-implicit.pdb')
forcefield = app.ForceField('amber99sb.xml', 'tip3p.xml')
system = forcefield.createSystem(pdb.topology, nonbondedMethod=app.NoCutoff)

psi = openmm.CustomTorsionForce('theta')
psi.addTorsion(6, 8, 14, 16)
psi_bias = app.BiasVariable(psi, -np.pi, np.pi, sigma, True, npoints)
meta = app.Metadynamics(system, [psi_bias], 300, 5, 1, nsteps)

integrator = openmm.VerletIntegrator(0.001)
platform = openmm.Platform.getPlatformByName('Reference')
simulation = app.Simulation(pdb.topology, system, integrator, platform)
simulation.context.setPositions(pdb.positions)
simulation.context.setVelocitiesToTemperature(300, 4321)
meta.step(simulation, nsteps)

cvforce = system.getForces()[-1]
x0 = cvforce.getCollectiveVariableValues(simulation.context)[0]
x = np.linspace(-np.pi, np.pi, num=npoints)
plt.plot(x, meta._totalBias, label='Metadynamics')
plt.plot(x, np.exp(-((x-x0)/sigma)**2/2), label='Exact')
plt.legend()
plt.show()
```
![Figure_1](https://user-images.githubusercontent.com/7718098/77688333-adc2ab00-6f7e-11ea-8c2b-7be0b8d60baf.png)

**Comment by craabreu** [comment]
`Metadynamics` fails to execute for a two-dimensional case if the grid size is not the same in the two directions. The script below reproduces this issue. The required PDB file is [here](https://github.com/openmm/openmm/blob/master/wrappers/python/tests/systems/alanine-dipeptide-implicit.pdb).

```py
from simtk import openmm
from simtk.openmm import app
import numpy as np 

pdb = app.PDBFile('alanine-dipeptide-implicit.pdb')
forcefield = app.ForceField('amber99sb.xml', 'tip3p.xml')
system = forcefield.createSystem(pdb.topology, nonbondedMethod=app.NoCutoff)

phi = openmm.CustomTorsionForce('theta')
phi.addTorsion(4, 6, 8, 14)
phi_bias = app.BiasVariable(phi, -np.pi, np.pi, 0.35, True, 101)

psi = openmm.CustomTorsionForce('theta')
psi.addTorsion(6, 8, 14, 16)
psi_bias = app.BiasVariable(psi, -np.pi, np.pi, 0.35, True, 181)

meta = app.Metadynamics(system, [phi_bias, psi_bias], 300, 5, 1, 1)

integrator = openmm.VerletIntegrator(0.001)
platform = openmm.Platform.getPlatformByName('Reference')
simulation = app.Simulation(pdb.topology, system, integrator, platform)
simulation.context.setPositions(pdb.positions)
simulation.context.setVelocitiesToTemperature(300, 4321)
meta.step(simulation, 1)
```

This is the resulting error traceback:

```
Traceback (most recent call last):
  File "test_metadynamics.py", line 24, in <module>
    meta.step(simulation, 1)
  File "/home/charlles/miniconda3/lib/python3.7/site-packages/simtk/openmm/app/metadynamics.py", line 169, in step
    self._addGaussian(position, height, simulation.context)
  File "/home/charlles/miniconda3/lib/python3.7/site-packages/simtk/openmm/app/metadynamics.py", line 211, in _addGaussian
    self._selfBias += height*gaussian
ValueError: operands could not be broadcast together with shapes (101,181) (181,101) (101,181) 
```

**Comment by peastman** [comment]
#2628 fixes the exception in the 2D case.  Can you try it and confirm it works for you?

**Comment by craabreu** [comment]
Thanks, @peastman. In fact, I have already solved this problem in #2611. The correct solution consists in changing the lines

https://github.com/openmm/openmm/blob/7c7dc7511446d05fe266c33389aea1db5d826a21/wrappers/python/simtk/openmm/app/metadynamics.py#L123-L124

into

```py
 self._selfBias = np.zeros(tuple(v.gridWidth for v in reversed(variables))) 
 self._totalBias = np.zeros(tuple(v.gridWidth for v in reserved(variables))) 
```

Otherwise, the data would be passed in the wrong order to `Continuous2DFunction` or  `Continuous3DFunction`.

**Comment by peastman** [comment]
Ok, thanks.
